# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/wr/RenderRootTypes.cpp
# Commit: 0c25260c4edf
# Full Hash: 0c25260c4edf97fdfef11320c2a91bc0698b4c11
# Author: Alexis Beingessner <a.beingessner@gmail.com>
# Date: 2019-07-25 03:53:31
# Regressor Bug: 1547351
# File Overlap Count: 2
# Description:
#   Bug 1547351 - Automatically set content process render root r=kats
#   
#   This implements the idea of automatically setting a content proc's
#   render root based on the render root enclosing the iframe that
#   points to it. There was a bit of cleanup in here that was a bit
# ==============================================================================

diff -r f96bd2aae39c -r 0c25260c4edf gfx/layers/wr/RenderRootTypes.cpp
--- a/gfx/layers/wr/RenderRootTypes.cpp	Wed Jul 24 14:10:56 2019 +0000
+++ b/gfx/layers/wr/RenderRootTypes.cpp	Wed Jul 24 14:10:56 2019 +0000
@@ -14,11 +14,13 @@
 void IPDLParamTraits<mozilla::layers::RenderRootDisplayListData>::Write(
     IPC::Message* aMsg, IProtocol* aActor, paramType&& aParam) {
   WriteIPDLParam(aMsg, aActor, aParam.mRenderRoot);
+  WriteIPDLParam(aMsg, aActor, aParam.mIdNamespace);
   WriteIPDLParam(aMsg, aActor, aParam.mRect);
   WriteIPDLParam(aMsg, aActor, aParam.mCommands);
   WriteIPDLParam(aMsg, aActor, aParam.mContentSize);
   WriteIPDLParam(aMsg, aActor, std::move(aParam.mDL));
   WriteIPDLParam(aMsg, aActor, aParam.mDLDesc);
+  WriteIPDLParam(aMsg, aActor, aParam.mRemotePipelineIds);
   WriteIPDLParam(aMsg, aActor, aParam.mResourceUpdates);
   WriteIPDLParam(aMsg, aActor, aParam.mSmallShmems);
   WriteIPDLParam(aMsg, aActor, std::move(aParam.mLargeShmems));
@@ -29,11 +31,13 @@
     const IPC::Message* aMsg, PickleIterator* aIter, IProtocol* aActor,
     paramType* aResult) {
   if (ReadIPDLParam(aMsg, aIter, aActor, &aResult->mRenderRoot) &&
+    ReadIPDLParam(aMsg, aIter, aActor, &aResult->mIdNamespace) &&
       ReadIPDLParam(aMsg, aIter, aActor, &aResult->mRect) &&
       ReadIPDLParam(aMsg, aIter, aActor, &aResult->mCommands) &&
       ReadIPDLParam(aMsg, aIter, aActor, &aResult->mContentSize) &&
       ReadIPDLParam(aMsg, aIter, aActor, &aResult->mDL) &&
       ReadIPDLParam(aMsg, aIter, aActor, &aResult->mDLDesc) &&
+      ReadIPDLParam(aMsg, aIter, aActor, &aResult->mRemotePipelineIds) &&
       ReadIPDLParam(aMsg, aIter, aActor, &aResult->mResourceUpdates) &&
       ReadIPDLParam(aMsg, aIter, aActor, &aResult->mSmallShmems) &&
       ReadIPDLParam(aMsg, aIter, aActor, &aResult->mLargeShmems) &&
@@ -43,6 +47,41 @@
   return false;
 }
 
+void WriteScrollUpdates(IPC::Message* aMsg, IProtocol* aActor,
+                        layers::ScrollUpdatesMap& aParam) {
+  // ICK: we need to manually serialize this map because
+  // nsDataHashTable doesn't support it (and other maps cause other issues)
+  WriteIPDLParam(aMsg, aActor, aParam.Count());
+  for (auto it = aParam.Iter(); !it.Done(); it.Next()) {
+    WriteIPDLParam(aMsg, aActor, it.Key());
+    WriteIPDLParam(aMsg, aActor, it.Data());
+  }
+}
+
+bool ReadScrollUpdates(const IPC::Message* aMsg, PickleIterator* aIter,
+                       IProtocol* aActor, layers::ScrollUpdatesMap* aResult) {
+  // Manually deserialize mScrollUpdates as a stream of K,V pairs
+  uint32_t count;
+  if (!ReadIPDLParam(aMsg, aIter, aActor, &count)) {
+    return false;
+  }
+
+  layers::ScrollUpdatesMap map(count);
+  for (size_t i = 0; i < count; ++i) {
+    layers::ScrollableLayerGuid::ViewID key;
+    layers::ScrollUpdateInfo data;
+    if (!ReadIPDLParam(aMsg, aIter, aActor, &key) ||
+        !ReadIPDLParam(aMsg, aIter, aActor, &data)) {
+      return false;
+    }
+    map.Put(key, data);
+  }
+
+  MOZ_RELEASE_ASSERT(map.Count() == count);
+  *aResult = std::move(map);
+  return true;
+}
+
 void IPDLParamTraits<mozilla::layers::RenderRootUpdates>::Write(
     IPC::Message* aMsg, IProtocol* aActor, paramType&& aParam) {
   WriteIPDLParam(aMsg, aActor, aParam.mRenderRoot);
@@ -50,7 +89,8 @@
   WriteIPDLParam(aMsg, aActor, aParam.mResourceUpdates);
   WriteIPDLParam(aMsg, aActor, aParam.mSmallShmems);
   WriteIPDLParam(aMsg, aActor, std::move(aParam.mLargeShmems));
-  WriteIPDLParam(aMsg, aActor, aParam.mScrollUpdates);
+  WriteScrollUpdates(aMsg, aActor, aParam.mScrollUpdates);
+  WriteIPDLParam(aMsg, aActor, aParam.mPaintSequenceNumber);
 }
 
 bool IPDLParamTraits<mozilla::layers::RenderRootUpdates>::Read(
@@ -61,7 +101,8 @@
       ReadIPDLParam(aMsg, aIter, aActor, &aResult->mResourceUpdates) &&
       ReadIPDLParam(aMsg, aIter, aActor, &aResult->mSmallShmems) &&
       ReadIPDLParam(aMsg, aIter, aActor, &aResult->mLargeShmems) &&
-      ReadIPDLParam(aMsg, aIter, aActor, &aResult->mScrollUpdates)) {
+      ReadScrollUpdates(aMsg, aIter, aActor, &aResult->mScrollUpdates) &&
+      ReadIPDLParam(aMsg, aIter, aActor, &aResult->mPaintSequenceNumber)) {
     return true;
   }
   return false;